File: pyalgorithm/array/contains_duplicate.py
import unittest
import collections

from typing import List

"""
    https://leetcode.com/problems/contains-duplicate/
    
    Cho mảng số nguyên, trả về true nếu có phần tử xuất hiện 2 lần
"""


class Solution(unittest.TestCase):

    def containsDuplicate(self, nums: List[int]) -> bool:
        counter = collections.Counter(nums)
        return any(v > 1 for v in counter.values())

    def test(self):
        self.assertEqual(True, self.containsDuplicate([1, 1, 2, 3, 4]))


if __name__ == '__main__':
    unittest.main()

File: leetcode/src/main/problems/array/ContainsDuplicate.java
package array;

import org.junit.jupiter.api.Test;

import java.util.*;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/contains-duplicate/
 *
 * Cho mảng số nguyên, trả về true nếu có phần tử xuất hiện 2 lần
 */
public class ContainsDuplicate {

    public boolean containsDuplicate0(int[] nums) {
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 1; i++) {
            if (nums[i] == nums[i + 1]) {
                return true;
            }
        }
        return false;
    }

    public boolean containsDuplicate(int[] nums) {
        Set<Integer> unique = new HashSet<>();
        for (int num : nums) {
            if (!unique.add(num)) {
                return true;
            }
        }
        return false;
    }

    @Test
    public void test() {
        assertEquals(true, containsDuplicate(new int[]{1, 2, 3, 1}));
        assertEquals(true, containsDuplicate(new int[]{1, 0, 1, 1}));
        assertEquals(true, containsDuplicate(new int[]{1, 2, 3, 1, 2, 3}));
        assertEquals(false, containsDuplicate(new int[]{1, 2, 3, 4, 5, 6}));
    }

}
